File: app/src/api/flashcard.ts
import { api } from "./api";
import type {
	UserFlashcard,
	UserFlashcardRequest,
	UserFlashcardStatusRequest,
	UserFlashcardWord,
	UserFlashcardWordRequest,
} from "./apiType";

// Flashcard
export async function listUserFlashcard(
	bookId: number,
): Promise<UserFlashcard[]> {
	try {
		const response = await api.get<UserFlashcard[]>(
			`/flashcard/book/${bookId}`,
		);
		if (!response.result || !response.data) return [];
		return response.data;
	} catch (error) {
		console.error("listUserFlashcard failed:", error);
		return [];
	}
}

export async function getUserFlashcard(
	flashcardId: number,
	cardType: string,
): Promise<UserFlashcard | undefined> {
	try {
		const response = await api.get<UserFlashcard>(
			`/flashcard/${flashcardId}/${cardType}`,
		);
		if (!response.result || !response.data) return undefined;
		return response.data;
	} catch (error) {
		console.error("getUserFlashcard failed:", error);
		return undefined;
	}
}

export async function createUserFlashcard(
	request: UserFlashcardRequest,
): Promise<UserFlashcard | null> {
	try {
		const response = await api.post<UserFlashcard>("/flashcard", request);
		if (!response.result || !response.data) return null;
		return response.data;
	} catch (error) {
		console.error("createUserFlashcard failed:", error);
		return null;
	}
}

// Flashcard Word
export async function listUserFlashcardWord(
	flashcardId: number,
): Promise<UserFlashcardWord[]> {
	try {
		const response = await api.get<UserFlashcardWord[]>(
			`/flashcard/word/${flashcardId}`,
		);
		if (!response.result || !response.data) return [];
		return response.data;
	} catch (error) {
		console.error("listUserFlashcardWord failed:", error);
		return [];
	}
}

export async function createUserFlashcardWord(
	request: UserFlashcardWordRequest,
): Promise<string | null> {
	try {
		const response = await api.post<string>("/flashcard/word", request);
		if (!response.result || !response.data) return null;
		return response.data;
	} catch (error) {
		console.error("createUserFlashcardWord failed:", error);
		return null;
	}
}

export async function upsertUserFlashcardWord(
	request: UserFlashcardWordRequest,
): Promise<UserFlashcardWord | null> {
	try {
		const response = await api.put<UserFlashcardWord>(
			"/flashcard/word",
			request,
		);
		if (!response.result || !response.data) return null;
		return response.data;
	} catch (error) {
		console.error("upsertUserFlashcardWord failed:", error);
		return null;
	}
}

export async function updateUserFlashcardStatus(
	request: UserFlashcardStatusRequest,
): Promise<UserFlashcard | null> {
	try {
		const response = await api.patch<UserFlashcard>(
			"/flashcard/status",
			request,
		);
		if (!response.result || !response.data) return null;
		return response.data;
	} catch (error) {
		console.error("updateUserFlashcardStatus failed:", error);
		return null;
	}
}

export async function deleteUserFlashcard(
	flashcardId: number,
	cardType: string,
): Promise<string | null> {
	try {
		const response = await api.delete<string>(
			`/flashcard/${flashcardId}/${cardType}`,
		);
		if (!response.result) return null;
		return "ok";
	} catch (error) {
		console.error("deleteUserFlashcard failed:", error);
		return null;
	}
}

export async function listUserFlashcardWordByType(
	flashcardId: number,
	cardType: string,
): Promise<UserFlashcardWord[]> {
	try {
		const response = await api.get<UserFlashcardWord[]>(
			`/flashcard/word/${flashcardId}/${cardType}`,
		);
		if (!response.result || !response.data) return [];
		return response.data;
	} catch (error) {
		console.error("listUserFlashcardWordByType failed:", error);
		return [];
	}
}

File: api/business/flashcard.py
from fastapi.encoders import jsonable_encoder
from accepter.base import UserFlashcardRequest, UserFlashcardWordRequest
from persistence.database import sessionScope
from persistence import repo_flashcard

async def listUserFlashcard(bookId: int, userId: str):
    with sessionScope() as db:
        encodedList = []
        flashcardList = await repo_flashcard.listUserFlashcard(userId, bookId, db)
        for flashcard in flashcardList :
            encodedList.append(jsonable_encoder(flashcard))

    return encodedList

async def getUserFlashcard(flashcardId: int, cardType: str, userId: str):
     with sessionScope() as db:
        flashCard = await repo_flashcard.getUserFlashcard(userId, flashcardId, cardType, db)        
        return jsonable_encoder(flashCard)

async def createUserFlashcard(req: UserFlashcardRequest, userId: str):
    with sessionScope() as db:
        flashCard = await repo_flashcard.createUserFlashcard(userId, 
                                                req.bookId, 
                                                req.flashcardId, 
                                                req.cardType,
                                                db)        
        return jsonable_encoder(flashCard)

async def listUserFlashcardWord(flashcardId: int, userId: str):
    with sessionScope() as db:
        encodedList = []
        wordList = await repo_flashcard.listUserFlashcardWord(userId, flashcardId, db)
        for word in wordList:
            encodedList.append(jsonable_encoder(word))

    return encodedList

async def createUserFlashcardWord(req: UserFlashcardWordRequest, userId: str):
    with sessionScope() as db:
        await repo_flashcard.createUserFlashcardWord(userId,
                                                req.flashcardId,
                                                req.cardType,
                                                req.cardId,
                                                req.status,
                                                db)
        return 'ok'

async def upsertUserFlashcardWord(req: UserFlashcardWordRequest, userId: str):
    with sessionScope() as db:
        word = await repo_flashcard.upsertUserFlashcardWord(userId,
                                                req.flashcardId,
                                                req.cardType,
                                                req.cardId,
                                                req.status,
                                                db)
        return jsonable_encoder(word)

async def updateUserFlashcardStatus(flashcardId: int, cardType: str, status: str, userId: str):
    with sessionScope() as db:
        result = await repo_flashcard.updateUserFlashcardStatus(userId, flashcardId, cardType, status, db)
        return jsonable_encoder(result)

async def deleteUserFlashcard(flashcardId: int, cardType: str, userId: str):
    with sessionScope() as db:
        await repo_flashcard.deleteUserFlashcard(userId, flashcardId, cardType, db)
        return 'ok'

async def listUserFlashcardWordByType(flashcardId: int, cardType: str, userId: str):
    with sessionScope() as db:
        encodedList = []
        wordList = await repo_flashcard.listUserFlashcardWordByType(userId, flashcardId, cardType, db)
        for word in wordList:
            encodedList.append(jsonable_encoder(word))
    return encodedList